slapo/schedule.py

```diff
@@ -903,29 +927,45 @@ def consolidate_model(sch: Schedule, topology=None):
             if is_found:
                 new_param = param.detach().split(sharded_size, dim=axis)[tp_rank]
                 sch.mod.register_parameter(param_name, nn.Parameter(new_param))
-    elif cnt_materialized != 0:
-        # skip if all the parameters have been materialized
-        assert (
-            cnt_meta == 0
-        ), f"Some of the parameters in module {sch.name} are on meta device"
-    elif cnt_meta == 0 and cnt_materialized == 0:
-        # skip if no direct parameters in this module
-        pass
-    for subsch in sch.child.values():
-        consolidate_model(subsch, topology)
+
+        for subsch in sch.child.values():
+            _consolidate_and_broadcast(subsch)
+
+    if cnt_meta != 0 or cnt_materialized != 0:
+        _consolidate_and_broadcast(sch)
+
+    gc.collect()
+    torch.cuda.empty_cache()
+
     return sch
 
 
-def build(sch: Schedule, topology=None, target=None, **kwargs):
+def build(
+    sch: Schedule,
+    topology=None,
+    target=None,
+    init_weights: Optional[Union[bool, Callable]] = True,
+    **kwargs,
+):
     optimizer = None
     if sch.metadata.pipeline_cutting_paths:
+        if topology is None:
+            raise ValueError("Must provide topology for deepspeed pipeline")
```

Comment:
We should not have this check at this position because target may not be "deepspeed".

Reply:
right. i will remove deepspeed pipeline